core/src/main/java/org/elasticsearch/ElasticsearchException.java

```diff
@@ -569,6 +562,23 @@ public static void generateFailureXContent(XContentBuilder builder, Params param
     }
 
     /**
+     * Parses the output of {@link #generateFailureXContent(XContentBuilder, Params, Exception, boolean)}
+     */
+    public static ElasticsearchException failureFromXContent(XContentParser parser) throws IOException {
+        XContentParser.Token token = parser.nextToken();
+        ensureFieldName(parser, token, ERROR);
+
+        token = parser.nextToken();
+        if (token.isValue() || token == XContentParser.Token.VALUE_NULL) {
+            return new ElasticsearchException(buildMessage("exception", parser.textOrNull(), null));
+        }
+
+        ensureExpectedToken(token, XContentParser.Token.START_OBJECT, parser::getTokenLocation);
```

Comment:
You're right, root causes are ignored because `innerFromXContent` ignores arrays of objects for now. I should have added a comment to make this clear, sorry.

I added a TODO for parsing root causes. I'd like to address this later in another issue because it might be tricky to parse them depending of the order of fields. Most of the time `root_causes` will come first but we must not rely on that and so `innerFromXContent` has to be adapted a bit.